### ISSUE TITLE
Recurse query doesn't return all nodes.

### ISSUE THREAD

**Issue by pawanrawal** [dgraph-io/dgraph#1753]
The issue is mentioned in detail in this post https://discuss.dgraph.io/t/multiple-variable-consistency/1932/5?u=pawan. I have been able to reproduce this. This seems a consequence of trying to avoid loops in the graph.

**Comment by pawanrawal** [comment]
The issue can be reproduced by doing the following mutation.
```
curl http://localhost:8080/mutate -H 'X-Dgraph-CommitNow: true'  -XPOST -d $'{
  set {
    <1> <name> "Alice" .
    <2> <name> "Bob" .
    <3> <name> "Charlie" .
    <1> <friend> <2> .
    <2> <friend> <3> .
  }
}'
```

After this if you do a query to get all friends like
```
curl localhost:8080/query -XPOST -d $'
{
  me(func: uid(1,2)) @recurse {
    friend
    name
  }
}'
```
The result doesn't have the relation between Bob <-> Charlie in the tree starting at Alice. This is because of how we detect cycles while doing recurse query.

**Comment by pawanrawal** [comment]
Fix has been merged to master 3b3af34d481aff641f8650e5bf8e0fa78e352144.
